docs: migrate 'Extensions' (#5454)
canonical/snapcraft: docs/explanation/extensions.rst

.. _explanation-extensions:

Extensions
==========

Snapcraft extensions enable snap developers to easily incorporate a set of common
requirements into a snap.

These requirements can include build and staging packages, plugs and interfaces, file
layouts and environments, and whatever other project file elements may be required to
build a functioning system.

A snap developer creating a GTK 3 application snap, for example, can use the
``gnome-3-28`` extension to expose the GTK 3 libraries to a snap at build and runtime
without the snap developer needing specific deep knowledge about GTK 3. There are
extensions for building robotics (ROS 2) applications too, including the :ref:`ROS2
Humble Extension <reference-ros-2-foxy-extension>`.

Extensions help:

* avoid repetitive tasks in the crafting process
* obviate the need for in-depth knowledge of the target software stack
* create a standard template for common application requirements
* reduce the testing and security burden, as they're tested and updated independently

> For a full list of supported extensions, see :ref:`reference-extensions`.

Reviewer: Do we have a precedence for mention the diataxis category?  Without it, this feels weird because you're already on a page about extensions.
```suggestion
For a full list of supported extensions, see the :ref:`reference-extensions` reference.
```

Author: There are a few other instances where we do this in the Snapcraft docs (see the end of the intro on [this page](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/how-to/crafting/configure-package-information/)).

I agree that it feels odd without it.

Author: https://github.com/canonical/snapcraft/pull/5454/commits/6d9b8bb17fa80be3e6c5abac012c4441e9fcfae0

Using extensions
----------------

> To use an extension, at least one :ref:`app <reference-snapcraft-yaml-app-keys>` in the

Reviewer: I feel like these sections overlap the extension how-tos.  There's already how-tos for using extensions and listing extensions.  I do like the sample output of `list-extensions` you added.  We are missing a how-to for the last step here, expanding extensions.

What do you think about adding a how-to for `expand-extensions` and dropping the rest of the content here?

Author: https://github.com/canonical/snapcraft/pull/5454/commits/6d9b8bb17fa80be3e6c5abac012c4441e9fcfae0

.. _explanation-extensions:

Extensions
==========

Snapcraft extensions enable snap developers to easily incorporate a set of common
requirements into a snap.

These requirements can include build and staging packages, plugs and interfaces, file
layouts and environments, and whatever other project file elements may be required to
build a functioning system.

> A snap developer creating a GTK 3 application snap, for example, can use the
> ``gnome-3-28`` extension to expose the GTK 3 libraries to a snap at build and runtime
> without the snap developer needing specific deep knowledge about GTK 3. There are
> extensions for building robotics (ROS 2) applications too, including the :ref:`ROS2
> Humble Extension <reference-ros-2-foxy-extension>`.

Reviewer 2: ```suggestion
A snap developer creating a GTK3 application snap, for example, can use the
gnome-3-28 extension to expose the GTK3 libraries to a snap at build and runtime
without the snap developer needing specific deep knowledge about GTK3. There are
extensions for building robotics (ROS 2) applications too, including the :ref:`ROS2
Humble Extension <reference-ros-2-foxy-extension>`.
```

Extensions help:

> * avoid repetitive tasks in the crafting process
> * obviate the need for in-depth knowledge of the target software stack
> * create a standard template for common application requirements
> * reduce the testing and security burden, as they're tested and updated independently

Reviewer 2: ```suggestion
- Avoid repetitive tasks in the crafting process
- Work around the need for in-depth knowledge of the target software stack
- Create a standard template for common application requirements
- Reduce the testing and security burden, as they're tested and updated independently
```
